packages/admin/admin/src/common/buttons/feedback/FeedbackButton.tsx

```diff
@@ -0,0 +1,181 @@
+import { ThreeDotSaving } from "@comet/admin-icons";
+import { Button as MuiButton, ButtonClassKey, ButtonProps, ComponentsOverrides } from "@mui/material";
+import { styled, Theme, useThemeProps } from "@mui/material/styles";
+import * as React from "react";
+
+import { Tooltip } from "../../Tooltip";
+
+export type FeedbackButtonClassKey = "idle" | "loading" | "success" | "fail" | ButtonClassKey;
+
+type OwnerState = Pick<FeedbackButtonProps, "variant" | "color"> & { displayState?: FeedbackButtonDisplayState };
+
+const Button = styled(MuiButton, {
+    name: "CometAdminFeedbackButton",
+    slot: "root",
+    overridesResolver({ ownerState }: { ownerState: OwnerState }, styles) {
+        return [
+            styles.root,
+            ownerState.displayState === "idle" && styles.idle,
+            ownerState.displayState === "fail" && styles.fail,
+            ownerState.displayState === "success" && styles.success,
+            ownerState.displayState === "loading" && styles.loading,
+        ];
+    },
+})<{ ownerState: OwnerState }>();
+
+export interface FeedbackButtonProps extends ButtonProps {
+    loading?: boolean;
+    hasErrors?: boolean;
+    loadingIcon?: React.ReactNode;
+    startIcon?: React.ReactNode;
+    endIcon?: React.ReactNode;
+    tooltipSuccessMessage?: string;
+    tooltipErrorMessage?: string;
+}
+
+type FeedbackButtonDisplayState = "idle" | "loading" | "success" | "fail";
+
+export function FeedbackButton(inProps: FeedbackButtonProps) {
+    const {
+        loading,
+        hasErrors,
+        children,
+        variant = "contained",
+        color = "primary",
+        classes,
+        disabled,
+        loadingIcon = <ThreeDotSaving />,
+        startIcon,
+        endIcon,
+        tooltipSuccessMessage,
+        tooltipErrorMessage,
+
+        ...restProps
+    } = useThemeProps({
+        props: inProps,
+        name: "CometAdminFeedbackButton",
+    });
+
+    const [displayState, setDisplayState] = React.useState<FeedbackButtonDisplayState>("idle");
+
+    const ownerState: OwnerState = {
+        displayState: displayState,
+        variant,
+        color,
+    };
+
+    const resolveIconForDisplayState = (displayState: FeedbackButtonDisplayState): any => {
+        if (startIcon && displayState !== "loading") {
+            return startIcon;
+        } else if (endIcon && displayState !== "loading") {
+            return endIcon;
+        }
+        if (displayState === "loading") {
+            return loadingIcon;
+        }
+    };
+
+    const resolveTooltipForDisplayState = (displayState: FeedbackButtonDisplayState): any => {
```

Comment:
Why explicitly set the return type to `any`. Can't the type simply be inferred?

```suggestion
    const resolveTooltipForDisplayState = (displayState: FeedbackButtonDisplayState) => {
```